src/main/java/games/mastermind/MMGameState.java

```diff
@@ -0,0 +1,140 @@
+package games.mastermind;
+
+import core.AbstractGameState;
+import core.AbstractParameters;
+import core.components.Component;
+import core.components.GridBoard;
+import core.components.PartialObservableDeck;
+import core.components.Token;
+import games.GameType;
+
+import java.util.*;
+
+import static games.mastermind.MMMethods.checkGuessAgainstAnswer;
+import static games.mastermind.MMMethods.generateRandomShuffledAnswerCode;
+
+public class MMGameState extends AbstractGameState {
+
+    PartialObservableDeck<Token> answerCode;
+    GridBoard<Token> guessBoard;
+    GridBoard<Token> resultBoard;
+    int activeRow;
+    int activeCol;
+
+    /**
+     * @param gameParameters - game parameters.
+     * @param nPlayers
+     */
+    public MMGameState(AbstractParameters gameParameters, int nPlayers) {
+        super(gameParameters, nPlayers);
+    }
+
+    @Override
+    protected GameType _getGameType() {
+        return GameType.Mastermind;
+    }
+
+    @Override
+    protected List<Component> _getAllComponents() {
+        return new ArrayList<>() {{
+            add(answerCode);
+            add(guessBoard);
+            add(resultBoard);
+        }};
+    }
+
+    @Override
+    protected AbstractGameState _copy(int playerId) {
+        MMParameters mmp = (MMParameters) getGameParameters();
+        MMGameState copy = new MMGameState(gameParameters, playerId);
+        copy.guessBoard = guessBoard.copy();
+        copy.resultBoard = resultBoard.copy();
+        copy.activeRow = activeRow;
+        copy.activeCol = activeCol;
+
+        copy.answerCode = answerCode.copy();
+        if (playerId != -1) {
+            PartialObservableDeck<Token> shuffledAnswerCode = generateRandomShuffledAnswerCode(mmp.boardWidth, redeterminisationRnd);
+            while (!satisfiesPreviousHints(shuffledAnswerCode)) {
+                shuffledAnswerCode = generateRandomShuffledAnswerCode(mmp.boardWidth, redeterminisationRnd);
+            }
+            copy.answerCode = shuffledAnswerCode;
+        }
+        return copy;
+    }
+
+    private boolean satisfiesPreviousHints(PartialObservableDeck<Token> answerCode) {
+        MMParameters mmp = (MMParameters) getGameParameters();
+        for (int i=0; i<activeRow-1; i++) {
+            List<Integer> result = checkGuessAgainstAnswer(this, answerCode, i);
+            for (int j=0; j<mmp.boardWidth; j++) {
+                if (!MMConstants.resultColours.get(result.get(j)).equals(resultBoard.getElement(j,i))) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    @Override
+    protected double _getHeuristicScore(int playerId) {
+        return getGameScore(playerId);
+    }
+
+    @Override
+    public double getGameScore(int playerId) {
```

Comment:
getGameScore should be reserved for an actual game score as per the rules of the game - which you can argue in Mastermind is equal to the number of rows unused. (Or, that there is no concept of score, in which case always return 0).

The code here is for a (very reasonable) heuristic.